solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java

```diff
@@ -174,20 +214,35 @@ public boolean hasPendingUpdates() {
    * @throws KeeperException       if any ZooKeeper operation results in an error
    * @throws InterruptedException  if the current thread is interrupted
    */
-  public ClusterState writePendingUpdates() throws IllegalStateException, KeeperException, InterruptedException {
+  public ClusterState writePendingUpdates(Map<String, ZkWriteCommand> updates) throws IllegalStateException, KeeperException, InterruptedException {
     if (invalidState) {
       throw new IllegalStateException("ZkStateWriter has seen a tragic error, this instance can no longer be used");
     }
-    if (!hasPendingUpdates()) return clusterState;
+    if ((updates == this.updates)
+        && !hasPendingUpdates()) {
+      return clusterState;
+    }
     Timer.Context timerContext = stats.time("update_state");
     boolean success = false;
     try {
       if (!updates.isEmpty()) {
-        for (Map.Entry<String, DocCollection> entry : updates.entrySet()) {
+        for (Map.Entry<String, ZkWriteCommand> entry : updates.entrySet()) {
           String name = entry.getKey();
           String path = ZkStateReader.getCollectionPath(name);
-          DocCollection c = entry.getValue();
+          ZkWriteCommand cmd = entry.getValue();
+          DocCollection c = cmd.collection;
+
+          if(cmd.ops != null && !cmd.ops.get().isEmpty() && cmd.ops.isPreOp()) {
+            PerReplicaStates.persist(cmd.ops, path, reader.getZkClient());
+            /*PerReplicaStates prs = PerReplicaStates.fetch(ZkStateReader.getCollectionPath(cmd.collection.getName()), reader.getZkClient());
+            log.debug("node {} ,per-replica states persisted {}->{}",reader.nodeName,
+                cmd.ops.getPerReplicaStates(),
+                prs);*/
```

Comment:
Delete commented out changes?

Reply:
These were left for anyone who wishes to debug something. I have added a `nocommit` will be removed right before merging the PR